lib/src/model/nameable.dart

```diff
@@ -29,8 +37,49 @@ abstract mixin class Nameable {
 
   @override
   String toString() => name;
+
+  PackageGraph get packageGraph;
+
+  /// Returns the [ModelElement] for [element], instantiating it if needed.
+  ModelElement getModelFor(
+    Element element,
+    Library library, {
+    Container? enclosingContainer,
+  }) =>
+      ModelElement.for_(
+        element,
+        library,
+        packageGraph,
+        enclosingContainer: enclosingContainer,
+      );
+
+  /// Returns the [ModelElement] for [element], instantiating it if needed.
+  ModelElement getModelForElement(Element element) =>
```

Comment:
What's the difference between `getModelForElement` and `getModelFor`? 

Additionally, could we not bundle this into the same helper with an optional field for `library`? So there's less choice in which to call. They all seem very similar.

Reply:
Asking the hard questions!

These three always trip me up, and I've just added a bit more documentation above `ModelElement.forElement`. I could add more documentation somewhere, but here's the story as I understand it:

* The most "naive" "generous" factory constructor is `forElement` (and this convenience method, `getModelForElement`. You can pass it an Element that is property-inducing, or not, and you don't even need to calculate the Library first.
* That factory constructor (1) determines the Library first, then (2) determines whether the element is property-inducing, and forks on that decision, calling either `for_` or `forPropertyInducingElement`.
* Now that we have two "delegatee" factory constructors, `for_` and `forPropertyInducingElement`, there are some callers that would prefer to call _them_, over `forElement`. Namely if you want to specify a specific enclosingContainer, or specific `getter`/`setter`.

Now, a lot of this is assumption: looking at the callers of `modelForPropertyInducingElement`, it looks like most callers calculate a `getter` and a `setter` in the same way that `forElement` does, and maybe that can all be collapsed. It would be great if I could eliminate all of the library-external calls to `modelForPropertyInducingElement`/`ModelElement.forPropertyInducingElement`.

It's also possible that we can strip the `library` parameter to these methods, if it is always the same object that would be determined by `ModelElement.forElement`. `ModelElement.forElement` uses a simple lookup. But maybe some callers need to construct a Library...

Also open to better names for these instead of relying wholly on doc comments to differentiate. ☹️

Reply:
> It would be great if I could eliminate all of the library-external calls to modelForPropertyInducingElement/ModelElement.forPropertyInducingElement.

Agreed.

Thanks for the documentation! Honestly, I think it's fine for now. Naming is hard.